posthog/temporal/batch_exports/s3_batch_export.py

```diff
@@ -970,23 +971,402 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+
+    async def consume_chunk(self, data: bytes):
+        if self._finalized:
+            raise RuntimeError("Consumer already finalized")
+
+        self.current_buffer.extend(data)
+        self.current_file_size += len(data)
+
+        # Upload parts when buffer is full
+        while len(self.current_buffer) >= self.part_size:
+            await self._upload_next_part()
+
+    async def _upload_next_part(self):
+        """Extract a part from buffer and upload it"""
+        if not self.upload_id:
+            await self._initialize_multipart_upload()
+
+        # Extract part data
+        part_data = bytes(self.current_buffer[: self.part_size])
+        self.current_buffer = self.current_buffer[self.part_size :]
+
+        part_number = self.part_counter
+        self.part_counter += 1
+
+        # Acquire memory semaphore (blocks if too many uploads in flight)
+        # await self.memory_semaphore.acquire()
+
+        # Create upload task
+        upload_task = asyncio.create_task(self._upload_part_with_cleanup(part_data, part_number))
+
+        # Track the upload
+        self.pending_uploads[part_number] = upload_task
+
+        # Clean up completed uploads to free memory
+        await self._cleanup_completed_uploads()
+
+    async def _upload_part_with_cleanup(self, data: bytes, part_number: int):
+        """Upload part and handle cleanup.
+
+        Note: This can run concurrently so need to be careful
+
+        TODO: add retry logic
+        """
+        # safety check - we should never have a part number without an upload id
+        if not self.upload_id:
+            raise NoUploadInProgressError()
+
+        try:
+            async with self.upload_semaphore:
+                await self.logger.adebug(
+                    "Uploading file number %s part %s with upload id %s",
+                    self.current_file_index,
+                    part_number,
+                    self.upload_id,
+                )
+                current_key = self._get_current_key()
+                async with self.s3_client() as client:
+                    response = await client.upload_part(
+                        Bucket=self.s3_inputs.bucket_name,
+                        Key=current_key,
+                        PartNumber=part_number,
+                        UploadId=self.upload_id,
+                        Body=data,
+                    )
+
+                part_info = {"ETag": response["ETag"], "PartNumber": part_number}
+
+                # Store completed part info
+                self.completed_parts[part_number] = part_info
+                return part_info
+
+        except Exception:
+            await self.logger.aexception(f"Failed to upload part {part_number}")
+            raise
+        finally:
+            # Always release memory semaphore
+            # self.memory_semaphore.release()
+            # Explicitly delete data to free memory immediately
+            del data
+            gc.collect()
```

Comment:
I am not entirely convinced whether this is the right tradeoff. Sure, the GC will run faster, but it means we'll run it a lot more (once per upload part). So, we are going to be GC blocked a lot more.